### PR TITLE
DF 2213 - delete data on unregistering user

### PR DESCRIPTION
**Pull request by emina-str**
- first draft

**TODO** : 
-- we have to trigger `JobAlertUnsubscribedEvent` on deleting SearchPofile, so the elasticsearch query from percolator is also deleted. --> https://alv-ch.atlassian.net/browse/DF-2276

Delete JobAdSearchProfile and JobAdFavouriteItem on UserUnregisteredEvent.

OS: https://github.com/alv-ch/alv-onlineform-service/pull/112
JOBROOM: https://github.com/alv-ch/jobroom2/pull/907
CANDIDATE: https://github.com/alv-ch/candidate-service/pull/170
JOBAD: https://github.com/alv-ch/job-ad-service/pull/442

### CODE REVIEW

File: application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/favouriteitem/FavouriteItemApplicationService.java [hunk 1]
```diff
@@ -116,6 +119,16 @@ public FavouriteItemDto findById(FavouriteItemId favouriteItemId) throws Favorit
                 .orElseThrow(() -> new FavoriteItemNotExitsException(favouriteItemId));
     }
 
+    @IsSysAdmin
+    public void handleUserUnregisteredEvent(String ownerUserId) {
+        notNull(ownerUserId, "OwnerUserId can't be null");
```
Comment (cjpsycos) [inline application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/favouriteitem/FavouriteItemApplicationService.java]: you should use Condition.notNull(ownerUserId, "OwnerUserId can't be null")

File: application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/searchprofile/SearchProfileApplicationService.java [hunk 2]
```diff
@@ -113,6 +115,16 @@ public void deleteSearchProfile(SearchProfileId searchProfileId) {
 		return new PageImpl<>(result, pageable, searchProfiles.getTotalElements());
 	}
 
+	@IsSysAdmin
+	public void handleUserUnregisteredEvent(String ownerUserId) {
+		notNull(ownerUserId, "OwnerUserId can't be null");
```
Comment (cjpsycos) [inline application/src/main/java/ch/admin/seco/jobs/services/jobadservice/application/searchprofile/SearchProfileApplicationService.java]: you should use Condition.notNull(ownerUserId, "OwnerUserId can't be null")

File: domain/src/main/java/ch/admin/seco/jobs/services/jobadservice/domain/favouriteitem/FavouriteItemRepository.java [hunk 3]
```diff
@@ -28,6 +28,10 @@
     @Query("select i from FavouriteItem i where i.jobAdvertisementId in :jobAdvertisementIds and i.ownerUserId = :ownerUserId")
     List<FavouriteItem> findByJobAdvertisementIdsAndOwnerId(@Param("jobAdvertisementIds") Set<JobAdvertisementId> jobAdvertisementIds, @Param("ownerUserId") String ownerUserId);
 
+    @Transactional(propagation = Propagation.SUPPORTS)
+    @Query("select i from FavouriteItem i where i.ownerUserId = :ownerUserId")
+    List<FavouriteItem> findByOwnerUserId(@Param("ownerUserId") String ownerUserId);
```
Comment (cjpsycos) [inline domain/src/main/java/ch/admin/seco/jobs/services/jobadservice/domain/favouriteitem/FavouriteItemRepository.java]: findAllByOwnerUserId

File: infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/users/UserEventDto.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package ch.admin.seco.jobs.services.jobadservice.infrastructure.messagebroker.users;
+
+public class UserEventDto {
+
+    private String userInfoId;
+
+    private String userInfoRole;
```
Comment (cjpsycos) [inline infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/users/UserEventDto.java]: only one role? user can have more than one role.

File: infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/LoginAsTechnicalUser.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package ch.admin.seco.jobs.services.jobadservice.infrastructure;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface LoginAsTechnicalUser {
```
Comment (cjpsycos) [inline infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/LoginAsTechnicalUser.java]: where is the imlementation of it?

File: infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/users/UserEventReceiverGateway.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    UserEventReceiverGateway(SearchProfileApplicationService searchProfileApplicationService, FavouriteItemApplicationService favouriteItemApplicationService) {
+        this.searchProfileApplicationService = searchProfileApplicationService;
+        this.favouriteItemApplicationService = favouriteItemApplicationService;
+    }
+
+    @StreamListener(value = MessageBrokerChannels.USER_EVENT_CHANNEL, condition = UNREGISTERED_EVENT_CONDITION)
+    @LoginAsTechnicalUser
+    public void handleUnregisteredEvent(UserEventDto event) {
+        LOGGER.info("Received UserUnregisteredEvent for User: '{}' with Role: '{}'.", event.getUserInfoId(), event.getUserInfoRole());
+        this.searchProfileApplicationService.handleUserUnregisteredEvent(event.getUserInfoId());
```
Comment (birom) [inline infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/users/UserEventReceiverGateway.java]: Same idea as here: https://github.com/alv-ch/candidate-service/pull/170/files#r395844897
I think the method names in the ``*ApplicationService`` should be specific. eg.: ``deleteUserSearchProfiles`` and ``deleteUserFavouriteItems``

File: infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/users/UserEventDto.java (hunk 4 above)
Comment (emina-str) [reply infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/messagebroker/users/UserEventDto.java]: it's a copy of event from Jobroom. There is only one role passed (either company, pav or jobseeker).

File: infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/LoginAsTechnicalUser.java (hunk 5 above)
Comment (emina-str) [reply infrastructure-messagebroker/src/main/java/ch/admin/seco/jobs/services/jobadservice/infrastructure/LoginAsTechnicalUser.java]: added :) thanks.

**Pull request merged**